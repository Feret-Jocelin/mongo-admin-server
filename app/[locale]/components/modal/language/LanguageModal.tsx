import { useTransition } from 'react';
import { useTranslations, useLocale } from 'next-intl';
import { usePathname, useRouter } from 'next-intl/client';

import GenericModal from "@components/modal/generic/GenericModal";

import { Language } from "@/domain/entities/setting-types";

import styles from "./languageModal.module.scss";

import { useDispatch } from "@/store/store";
import { setLanguage } from "@/domain/usecases/setting-slice";

interface LanguageModalProps {
  open: boolean;
  onClose: () => void;
}

const LanguageModal = ({ open, onClose }: LanguageModalProps) => {
  const router = useRouter();
  const t = useTranslations();
  const locale = useLocale();
  const [isPending, startTransition] = useTransition();
  const pathname = usePathname();
  const dispatch = useDispatch();

  const languages: Language[] = [
    { value: 'fr', label: t('language.fr'), },
    { value: 'en', label: t('language.en'), },
    { value: 'es', label: t('language.es'), },
  ];

  const handleLanguageChange = (language: string) => {
    startTransition(() => {
      router.replace(pathname, {locale: language});
    });
    dispatch(setLanguage(language))
    onClose();
  };
  
  return (
    <GenericModal
      open={open}
      title={t('modal.language.title')}
      withButton={false}
      onClose={onClose}
    >
      <ul className={styles["language-list"]}>
      {languages.map((language, index) => (
        <li 
          key={index} 
          className={`${styles["language-container"]} ${language.value === locale ? styles['selected'] : ''}`}
          onClick={() => handleLanguageChange(language.value)}
        >
          {language.label}
        </li>
      ))}
      </ul>
    </GenericModal>
  );
}

export default LanguageModal;
