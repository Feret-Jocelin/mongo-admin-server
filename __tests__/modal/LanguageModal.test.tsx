/* eslint-disable no-undef */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { NextIntlClientProvider } from 'next-intl';
import { store } from '@/store/store';
import { Provider } from 'react-redux';
import LanguageModal from '@/app/[locale]/components/modal/language/LanguageModal';

jest.mock('next-intl/client', () => ({
  usePathname() {
    return '';
  },
  useRouter() {
    return {
      push: () => jest.fn(),
      replace: () => jest.fn(),
    };
  },
}));

describe("LanguageModal", () => {
  const locale = 'fr';
  const message = require(`../../shared/locales/${locale}.json`)
  it("renders", () => {
    const handleClose = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <NextIntlClientProvider locale={locale} messages={message}>
          <LanguageModal 
            open={false}
            onClose={handleClose}
          />
        </NextIntlClientProvider>
      </Provider>
    );
    const title = screen.getByText('Langue');
    expect(title).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  })
})