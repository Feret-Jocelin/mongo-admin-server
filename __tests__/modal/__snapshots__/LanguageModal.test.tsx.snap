// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageModal renders 1`] = `
<div>
  <dialog
    class="modal modal--closing"
  >
    <section
      class="modal__header"
    >
      <div
        class="modal__header--icon"
      >
        <div
          class="undefined pointer-cursor"
        />
      </div>
      <h2>
        Langue
      </h2>
    </section>
    <section
      class="modal__container"
    >
      <ul
        class="language-list"
      >
        <li
          class="language-container selected"
        >
          ğŸ‡«ğŸ‡· FranÃ§ais
        </li>
        <li
          class="language-container "
        >
          ğŸ‡¬ğŸ‡§ Anglais
        </li>
        <li
          class="language-container "
        >
          ğŸ‡ªğŸ‡¸ Espagnol
        </li>
      </ul>
    </section>
  </dialog>
</div>
`;
